File: cpp/Kanzi/io/InfoPrinter.cpp
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <sstream>
#include <cstdio>
#include "InfoPrinter.hpp"

using namespace kanzi;

InfoPrinter::InfoPrinter(int infoLevel, InfoPrinter::Type type, OutputStream& os)
    : _os(os)
{
    _level = infoLevel;
    _type = type;

    if (type == InfoPrinter::ENCODING) {
        _thresholds[0] = BlockEvent::BEFORE_TRANSFORM;
        _thresholds[1] = BlockEvent::AFTER_TRANSFORM;
        _thresholds[2] = BlockEvent::BEFORE_ENTROPY;
        _thresholds[3] = BlockEvent::AFTER_ENTROPY;
    }
    else {
        _thresholds[0] = BlockEvent::BEFORE_ENTROPY;
        _thresholds[1] = BlockEvent::AFTER_ENTROPY;
        _thresholds[2] = BlockEvent::BEFORE_TRANSFORM;
        _thresholds[3] = BlockEvent::AFTER_TRANSFORM;
    }
}

void InfoPrinter::processEvent(const BlockEvent& evt)
{
    int currentBlockId = evt.getId();

    if (evt.getType() == _thresholds[0]) {
        // Register initial block size
        BlockInfo* bi = new BlockInfo();
        bi->time0 = evt.getTime();

        if (_type == InfoPrinter::ENCODING)
            bi->stage0Size = evt.getSize();

        _map.insert(pair<int, BlockInfo*>(currentBlockId, bi));

        if (_level >= 4) {
            _os << evt.toString() << endl;
        }
    }
    else if (evt.getType() == _thresholds[1]) {
        map<int, BlockInfo*>::iterator it = _map.find(currentBlockId);

        if (it == _map.end())
            return;

        BlockInfo* bi = it->second;
        bi->time1 = evt.getTime();

        if (_type == InfoPrinter::DECODING)
            bi->stage0Size = evt.getSize();

        if (_level >= 4) {
            int duration_ms = int(1000 * (bi->time1 - bi->time0) / CLOCKS_PER_SEC);
            stringstream ss;
            ss << evt.toString() << " [" << duration_ms << " ms]";
            _os << ss.str() << endl;
        }
    }
    else if (evt.getType() == _thresholds[2]) {
        map<int, BlockInfo*>::iterator it = _map.find(currentBlockId);

        if (it == _map.end())
            return;

        BlockInfo* bi = it->second;
        bi->time2 = evt.getTime();
        bi->stage1Size = evt.getSize();

        if (_level >= 4) {
            int duration_ms = int(1000 * (bi->time2 - bi->time1) / CLOCKS_PER_SEC);
            stringstream ss;
            ss << evt.toString() << " [" << duration_ms << " ms]";
            _os << ss.str() << endl;
        }
    }
    else if (evt.getType() == _thresholds[3]) {
        int stage2Size = evt.getSize();
        map<int, BlockInfo*>::iterator it = _map.find(currentBlockId);

        if (it == _map.end())
            return;

        if (_level < 2) {
            delete it->second;
            _map.erase(it);
            return;
        }

        BlockInfo* bi = it->second;
        bi->time3 = evt.getTime();
        int duration1_ms = int(1000 * (bi->time1 - bi->time0) / CLOCKS_PER_SEC);
        int duration2_ms = int(1000 * (bi->time3 - bi->time2) / CLOCKS_PER_SEC);
        stringstream ss;

        if (_level >= 4) {
            ss << evt.toString() << " [" << duration2_ms << " ms]" << endl;
        }

        // Display block info
        if (_level >= 3) {
            ss << "Block " << currentBlockId << ": " << bi->stage0Size << " => ";
            ss << bi->stage1Size << " [" << duration1_ms << " ms] => " << stage2Size;
            ss << " [" << duration2_ms << " ms]";
        }
        else {
            ss << "Block " << currentBlockId << ": " << bi->stage0Size << " => ";
            ss << bi->stage1Size << " => " << stage2Size;
        }

        // Add compression ratio for encoding
        if (_type == InfoPrinter::ENCODING) {
            if (bi->stage0Size != 0) {
                char buf[32];
                sprintf(buf, " (%d%%)", (int)(stage2Size * (double)100 / (double)bi->stage0Size));
                ss << buf;
            }
        }

        // Optionally add hash
        if (evt.getHash() != 0) {
            char buf[32];
            sprintf(buf, " [%08X]", evt.getHash());
            ss << buf;
        }

        _os << ss.str() << endl;
        delete bi;
        _map.erase(it);
    }
}

File: java/src/main/java/kanzi/app/InfoPrinter.java
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.app;

import kanzi.Event;
import java.io.PrintStream;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import kanzi.Listener;

// An implementation of Listener to display block information (verbose option
// of the BlockCompressor/BlockDecompressor)
public class InfoPrinter implements Listener
{
   public enum Type { ENCODING, DECODING }
   
   private final PrintStream ps;
   private final Map<Integer, BlockInfo> map;
   private final Event.Type[] thresholds;
   private final Type type;
   private final int level;
   
   
   public InfoPrinter(int infoLevel, Type type, PrintStream ps) 
   {
      if (ps == null)
         throw new NullPointerException("Invalid null print stream parameter");
      
      this.ps = ps;
      this.level = infoLevel;
      this.type = type;
      this.map = new ConcurrentHashMap<Integer, BlockInfo>();
      this.thresholds = (type == Type.ENCODING) ? 
              new Event.Type[]
              { 
                 Event.Type.COMPRESSION_START,
                 Event.Type.BEFORE_TRANSFORM,
                 Event.Type.AFTER_TRANSFORM,
                 Event.Type.BEFORE_ENTROPY,
                 Event.Type.AFTER_ENTROPY,
                 Event.Type.COMPRESSION_END
              } :
              new Event.Type[]
              { 
                 Event.Type.DECOMPRESSION_START,
                 Event.Type.BEFORE_ENTROPY,
                 Event.Type.AFTER_ENTROPY,
                 Event.Type.BEFORE_TRANSFORM,
                 Event.Type.AFTER_TRANSFORM,
                 Event.Type.DECOMPRESSION_END
              };
   }
   
   
   @Override
   public void processEvent(Event evt) 
   {
      int currentBlockId = evt.getId();

      if (evt.getType() == this.thresholds[1])
      {
         // Register initial block size
         BlockInfo bi = new BlockInfo();
         bi.time0 = evt.getTime();
         
         if (this.type == Type.ENCODING)
            bi.stage0Size = evt.getSize();
         
         this.map.put(currentBlockId, bi);
         
         if (this.level >= 5)
         {
            this.ps.println(evt);
         }
      }
      else if (evt.getType() == this.thresholds[2])
      { 
         BlockInfo bi = this.map.get(currentBlockId);
         
         if (bi == null)
            return;
         
         bi.time1 = evt.getTime();
         
         if (this.type == Type.DECODING)
            bi.stage0Size = evt.getSize();
                          
         if (this.level >= 5)
         {
            long duration_ms = (bi.time1 - bi.time0) / 1000000L; 
            this.ps.println(String.format("%s [%d ms]", evt, duration_ms));
         }
      }
      else if (evt.getType() == this.thresholds[3])
      {
         BlockInfo bi = this.map.get(currentBlockId);
         
         if (bi == null)
            return;
         
         bi.time2 = evt.getTime();
         bi.stage1Size = evt.getSize();
         
         if (this.level >= 5)
         {
            long duration_ms = (bi.time2 - bi.time1) / 1000000L; 
            this.ps.println(String.format("%s [%d ms]", evt, duration_ms));
         }
      }
      else if (evt.getType() == this.thresholds[4])
      {        
         long stage2Size = evt.getSize();
         BlockInfo bi = this.map.remove(currentBlockId);
         
         if ((bi == null) || (this.level < 3))
            return;
             
         bi.time3 = evt.getTime();
         long duration1_ms = (bi.time1 - bi.time0) / 1000000L; 
         long duration2_ms = (bi.time3 - bi.time2) / 1000000L; 
         StringBuilder msg = new StringBuilder();
                  
         if (this.level >= 5)
         {
            this.ps.println(String.format("%s [%d ms]", evt, duration2_ms));
         }
         
         // Display block info
         if (this.level >= 4)
         {
            msg.append(String.format("Block %d: %d => %d [%d ms] => %d [%d ms]", currentBlockId, 
                    bi.stage0Size, bi.stage1Size, duration1_ms, stage2Size, duration2_ms));

            // Add compression ratio for encoding
            if (this.type == Type.ENCODING)
            {
               if (bi.stage0Size != 0)
                  msg.append(String.format(" (%d%%)", (stage2Size*100L/(long) bi.stage0Size)));
            }

            // Optionally add hash
            if (evt.getHash() != null) 
               msg.append(String.format("  [%s]", Integer.toHexString(evt.getHash())));

            this.ps.println(msg.toString());         
         }
      }
      else if ((evt.getType() == Event.Type.AFTER_HEADER_DECODING) && (this.level >= 3))
      {
         this.ps.println(evt);
      }
      else if (this.level >= 5)
      {
         this.ps.println(evt);
      }
   }
   
   
   static class BlockInfo
   {
      long time0;
      long time1;
      long time2;
      long time3;
      long stage0Size;
      long stage1Size;
   }
}
